pinns/src/pinns.c

```diff
@@ -268,20 +294,62 @@ static bool is_host_ns(const char* const optarg) {
   return optarg && !strcmp (optarg, HOSTNS);
 }
 
+// Verify the /$PATH/$NSns directory exists and is a directory, returning the
+// expanded path name in bind_root and returning the length of bind_root (or -1
+// on error)
+static int create_bind_root(char *bind_root, size_t size, const char *pin_path, const char *ns_name) {
+  int len = snprintf(bind_root, size, "%s/%sns", pin_path, ns_name);
+  if (directory_exists_or_create(bind_root) < 0) {
+    pwarnf("%s exists and is not a directory", bind_root);
+    return -1;
+  }
+  return len;
+}
+
+/* Mount namespaces can only be bound into unsharable mount namespaces (to
+ * avoid infinite loops), so force the pin_path to be a bind-mount to
+ * itself and then marked as unsharable. */
+static int setup_unbindable_bindpath(const char *pin_path, const char *ns_name) {
+  char bind_root[PATH_MAX];
+
+  // first, verify the /$PATH/$NSns directory exists
+  if (create_bind_root(bind_root, PATH_MAX - 1, pin_path, ns_name) < 0) {
+    return -1;
+  }
+
+  // TODO: Check if bind_root is already a mountpoint
+  // For now, just blindly try to bindmount itself to itself.
+  // If this succeeds, we know it's a mountpoint.  If it fails it probably
+  // failed because it's already a mountpoint, and if it's not, the call to set
+  // MS_UNBINDABLE will fail next.
+  if (mount(bind_root, bind_root, NULL, MS_BIND, NULL) < 0) {
```

Comment:
it seems we're not reacting to mount()<0. Why have the if statement?

Reply:
2 reasons: A lot of coverage tools get cranky if you don't check the return value in some way, and I wanted to make it explicit that we're explicitly ignoring errors.

Happy to do it some other way.

Reply:
Trying to just ignore it to see if any tooling here complains...